File: Sigmoid.lua
local Sigmoid = torch.class('nn.Sigmoid', 'nn.Module')

function Sigmoid:updateOutput(input)
   local start=sys.clock()
   input.THNN.Sigmoid_updateOutput(
      input:cdata(),
      self.output:cdata()
   )
   sys.Sigmoid_F = sys.Sigmoid_F + sys.clock() - start

   return self.output
end

function Sigmoid:updateGradInput(input, gradOutput)
   local start=sys.clock()
   input.THNN.Sigmoid_updateGradInput(
      input:cdata(),
      gradOutput:cdata(),
      self.gradInput:cdata(),
      self.output:cdata()
   )
   sys.Sigmoid_B = sys.Sigmoid_B + sys.clock() - start
   return self.gradInput
end

File: lib/THNN/generic/Sigmoid.c
#ifndef TH_GENERIC_FILE
#define TH_GENERIC_FILE "generic/Sigmoid.c"
#else

void THNN_(Sigmoid_updateOutput)(
          THNNState *state,
          THTensor *input,
          THTensor *output)
{
  THTensor_(sigmoid)(output, input);
}

void THNN_(Sigmoid_updateGradInput)(
          THNNState *state,
          THTensor *input,
          THTensor *gradOutput,
          THTensor *gradInput,
          THTensor *output)
{
  THNN_CHECK_NELEMENT(output, gradOutput);
  THTensor_(resizeAs)(gradInput, output);
  ptrdiff_t gradInputSize = THTensor_(nElement)(gradInput);                     
  ptrdiff_t gradOutputSize = THTensor_(nElement)(gradOutput);   
  ptrdiff_t outputSize = THTensor_(nElement)(output);                    
  int gradInputContig = THTensor_(isContiguous)(gradInput)? 1:0;                 
  int gradOutputContig = THTensor_(isContiguous)(gradOutput)? 1:0;   
  int outputContig = THTensor_(isContiguous)(output)? 1:0; 
  if((gradInputSize = gradOutputSize) && (gradInputSize == outputSize) ){	
    TH_TENSOR_APPLY3_ADVANCED_INDEX2(outputSize, gradInputContig, gradOutputContig, outputContig,
                                     real, gradInput, real, gradOutput, real, output,
                                     real z = *output_data;
                                     *gradInput_data = *gradOutput_data * (1. - z) * z;
    );
  } else {
    TH_TENSOR_APPLY3(real, gradInput, real, gradOutput, real, output,
      real z = *output_data;
      *gradInput_data = *gradOutput_data * (1. - z) * z;
    );
	  
  }
}

#endif
